_posts/2021-12-08-kafkasql-storage-and-security.md

```diff
@@ -0,0 +1,471 @@
+
+The design of the Kafka messages closely follows the tables that are used both in the local SQL database, 
+and the database for the standard SQL storage.
+
+Each Kafka message has a key and a corresponding value. Kafka itself is agnostic about their content, and treats them as a byte array.
+In the KafkaSQL storage, they are represented by objects that are serialized into a JSON String, 
+prepended with marker bytes used to recognize message type. 
+
+In general, each message value class has a corresponding key class. 
+Classes for the keys are mainly responsible for determining the partition that the message is going to reside in, 
+the tenant information in case of a multi-tenant deployment, and identifying information for the object that the message value represents. 
+For example `ArtifactVersionKey` contains group ID, artifact ID and a version identifiers for the given artifact version object.
+Choice of the partition key is important for maintaining consistency of the data between nodes, 
+because ordering of the messages is only guaranteed in the same partition.
+
+Classes fo the values contain a field with operation that the message represents, which includes `CREATE`, `UPDATE`, and `DELETE`.
+The other fields carry the rest of the information, such as the editable metadata for an artifact version.
+
+![KafkaSQL Storage Diagram](/images/guides/registry-storage-kafkasql-diagram.svg)
+*KafkaSQL storage option workflow diagram*
+
+When a change to a piece of data occurs, the originating nodes creates and sends the message to the Kafka topic,
+and blocks the user thread until the operation is processed. 
+Its local database is not modified directly. Each node only applies changes that are read form the Kafka topic.
+Therefore, user thread is unblocked only after the original message is read back from the topic by the node.
+This ensures that in case of a composite operation, messages are always processed in the correct order.
+
+For example, when adding a new artifact version, the user may store a new content that is not already
+present in the storage. Therefore, the operation usually consists of two messages that have to be kept in the correct order,
+or the artifact version wouldn't be able to reference the new content.
+
+An important thing to note is that when a node is restarted, and its local SQL database is empty,
+simply reading all the messages from the `kafkasql-journal` will replicate the data that is already present on other live nodes.
+This works even when the entire Apicurio Registry application is stopped and restarted,
+as long as the data in the `kafkasql-journal` topic remains persisted and backed up.
+
+# Investigation
+
+In order to investigate the issue, we have to first attempt to reproduce the issue on a local Apicurio Registry deployment.
+The following is a list of steps we have performed.
```

Comment:
I don't think there is an explanation of the issue.  What exactly is the issue?  In the intro it mentions a bug, but doesn't describe what the bug is.

Reply:
I do think a section at the top (before describing KafkaSQL storage perhaps) is needed to just explain what users were experiencing.  Mentioning that the issue was only a problem in the KafkaSQL storage, which should be a nice lead-in to that section.